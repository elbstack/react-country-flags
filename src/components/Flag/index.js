import React, { PureComponent as Component, createElement } from 'react';
import PropTypes from 'prop-types';
import flags from '../../flags/react';

const fetchFlag = async country => {
  if (!flags[country]) return Promise.resolve(null);
  const { default: data } = await flags[country]();
  return data;
};

// thanks to https://github.com/facebook/react/issues/5465#issuecomment-157888325
const makeCancelable = promise => {
  let hasCanceled_ = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    promise.then(
      val => (hasCanceled_ ? reject({ isCanceled: true }) : resolve(val)),
      error => (hasCanceled_ ? reject({ isCanceled: true }) : reject(error))
    );
  });

  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled_ = true;
    },
  };
};

export const countries = (() => Object.keys(flags))();

export default class Flags extends Component {
  cancelUpdate = () => {};

  state = {
    flag: null,
  };

  static defaultProps = {
    country: null,
    asSquare: false,
  };

  componentDidMount() {
    this.updateFlag();
  }

  componentDidUpdate(prevProps) {
    const { country } = this.props;

    if (country !== prevProps.country) {
      this.updateFlag();
    }
  }

  componentWillUnmount() {
    this.cancelUpdate();
  }

  updateFlag() {
    const { country } = this.props;
    const { promise, cancel } = makeCancelable(fetchFlag(country));
    promise.then(flag => {
      this.setState({ flag });
    });
    this.cancelUpdate = cancel;
  }

  render() {
    const { asSquare, country, ...props } = this.props;
    const { flag } = this.state;
    if (!flag && !asSquare) return null;
    if (!asSquare) return createElement(flag, props);
    return (
      <svg viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" {...props}>
        {!flag ? null : createElement(flag)}
      </svg>
    );
  }
}

Flags.propTypes = {
  country: PropTypes.string,
  asSquare: PropTypes.bool,
};
