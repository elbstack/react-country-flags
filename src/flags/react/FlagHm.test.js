import React from 'react';
import renderer from 'react-test-renderer';
import Flag from './FlagHm';

it('FlagHm renders without crashing', () => {
  const component = renderer.create(<Flag />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
