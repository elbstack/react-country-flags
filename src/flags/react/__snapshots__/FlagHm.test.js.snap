// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlagHm renders without crashing 1`] = `
<svg
  viewBox="0 0 10080 5040"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <defs>
    <clipPath
      id="s061baa09"
    >
      <path
        d="M0 0v1.5h7V3H6zm6 0H3v3.5H0V3z"
      />
    </clipPath>
    <path
      d="M0-360l69.421 215.845 212.038-80.301L155.99-35.603l194.985 115.71-225.881 19.651 31.105 224.59L0 160l-156.198 164.349 31.105-224.59-225.881-19.651 194.986-115.711-125.471-188.853 212.038 80.301z"
      id="sa5fd8f2c"
    />
    <path
      d="M0-210L54.86-75.508l144.862 10.614L88.765 28.842l34.67 141.052L0 93.334l-123.435 76.56 34.67-141.052-110.957-93.736L-54.86-75.508z"
      id="s39263a8a"
    />
    <use
      id="s26848e82"
      transform="scale(2.1)"
      xlinkHref="#sa5fd8f2c"
    />
  </defs>
  <path
    d="M0 0h10080v5040H0z"
    fill="#00008b"
  />
  <path
    d="M0 0l5040 2520m0-2520L0 2520"
    stroke="#fff"
    strokeWidth={504}
  />
  <path
    clipPath="url(#s061baa09)"
    d="M0 0l6 3m0-3L0 3"
    stroke="red"
    strokeWidth={0.4}
    transform="scale(840)"
  />
  <path
    d="M2520 0v2940M0 1260h5880"
    stroke="#fff"
    strokeWidth={840}
  />
  <path
    d="M2520 0v2940M0 1260h5880"
    stroke="red"
    strokeWidth={504}
  />
  <path
    d="M0 2520h5040V0h1680v3360H0z"
    fill="#00008b"
  />
  <g
    fill="#fff"
  >
    <use
      x={2520}
      xlinkHref="#s26848e82"
      y={3780}
    />
    <use
      x={7560}
      xlinkHref="#sa5fd8f2c"
      y={4200}
    />
    <use
      x={6300}
      xlinkHref="#sa5fd8f2c"
      y={2205}
    />
    <use
      x={7560}
      xlinkHref="#sa5fd8f2c"
      y={840}
    />
    <use
      x={8680}
      xlinkHref="#sa5fd8f2c"
      y={1869}
    />
    <use
      x={8064}
      xlinkHref="#s39263a8a"
      y={2730}
    />
  </g>
</svg>
`;
